### ISSUE TITLE
[Twig] Enable autoscaping on {% trans with { .. vars .. } %}

### ISSUE THREAD

**Issue by ruudk** [symfony/symfony#21565]
| Q                | A
| ---------------- | -----
| Bug report?      | no
| Feature request? | yes
| BC Break report? | no
| RFC?             | no
| Symfony version  | 3.2

## The problem
Our project heavily uses `{% trans %}key{% endtrans %}` instead of the `{{ key | trans }}` filter. The reason for this is that we sometimes have HTML tags in our translation keys. 

This works perfectly. The only downside if that we sometimes do stuff like this:
```yaml
my.translation.key: 'Welcome to %event%'
```

```twig
{% trans with { 
    '%event%': event.title
} %}my.translation.key{% endtrans %}
```

You probably already see the issue here: XSS injection possibilities! To fix this, simply add ` | escape` to the passed variables like this:
```twig
{% trans with { 
    '%event%': event.title | trans
} %}my.translation.key{% endtrans %}
```

But this is something that is easily forgotten. Especially since Twig is so awesome in auto escaping everything that developers don't have to think about it.

## The possible solution

Would it be possible if the passed variables to `{% trans with {} %}` would always be escaped automatically, unless you specify ` | raw` ?

**Comment by stof** [comment]
Escaping the placeholder value alone may not be enough to forbid XSS. It depends how it is used inside the translation string.
And anyway, the Twig autoescaping always escape when outputting something. Here, the ``event.title`` is not output, but passed to a tag (which may output it internally, but Twig cannot know it).
If we turned on autoescaping, it would apply on the output of the ``{% trans %}`` tag (but this would be a BC break).
Btw, this is the reason I never use this tag but always the filter. This way, I'm safe by default, and I must explicitly opt-in to allow HTML tags in translation strings (in which case, I will also escape the necessary things, as I know I need to be careful when I start using ``|raw`` on the output)

**Comment by ruudk** [comment]
@stof Thanks for your fast reply. I was already thinking it wasn't as simple as I thought.

I decided to create a CLI command that tokenises every Twig template and then parses it into a node tree. It then searches the tree to find `TransNode`. If the `TransNode` has `vars` it will then do some checking to see if the value was `Twig_Node_Expression_Constant` or dynamic `Twig_Node_Expression_Filter`. It then checks if the expression contains a `raw` or `escape` filter, and will fail otherwise.

This way I found 11 potential XSS issues in our project 😅
